src/libusermode/userhook.cpp

```diff
@@ -373,58 +353,103 @@ static event_response_t perform_hooking(drakvuf_t drakvuf, drakvuf_trap_info* in
     return ret;
 }
 
-event_response_t hook_dll(drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t base_address)
+event_response_t hook_dll(drakvuf_t drakvuf, drakvuf_trap_info_t* info, mmvad_info_t* mmvad, bool* is_hooked)
 {
     auto plugin = get_trap_plugin<userhook>(info);
 
     dll_t* dll_meta = get_pending_dll(drakvuf, info, plugin);
 
     if (!dll_meta)
-        dll_meta = create_dll_meta(drakvuf, info, plugin, base_address);
+        dll_meta = create_dll_meta(drakvuf, info, plugin, mmvad);
+
+    event_response_t res = VMI_EVENT_RESPONSE_NONE;
+
+    if (is_hooked)
+        *is_hooked = false;
 
     if (dll_meta)
-        return perform_hooking(drakvuf, info, plugin, dll_meta);
+    {
+        res = perform_hooking(drakvuf, info, plugin, dll_meta);
+        if (is_hooked)
+            *is_hooked = dll_meta->v.is_hooked;
+    }
 
-    return VMI_EVENT_RESPONSE_NONE;
+    return res;
 }
 
 static void search_process_system_dlls(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
 {
     auto plugin = get_trap_plugin<userhook>(info);
 
-    struct visitor_context
+    struct visitor_context_t
     {
-        drakvuf_trap_info_t* info{nullptr};
-        userhook* plugin{nullptr};
-
-        addr_t dll_base{0};
+        std::string name;
+        mmvad_info_t mmvad;
     };
 
-    if (auto search = plugin->proc_ntdll_hooked.find(info->attached_proc_data.pid);
-        search == plugin->proc_ntdll_hooked.end() || !search->second)
+    userhook::module_context_t* ctx = nullptr;
+
+    auto it = plugin->proc_ntdll.find(info->attached_proc_data.pid);
+    if (it != plugin->proc_ntdll.end())
     {
-        auto visitor = [](drakvuf_t drakvuf_, const module_info_t* module_info, bool* need_free, bool* need_stop, void* ctx)
+        ctx = &it->second;
+    }
+    else
+    {
+        auto visitor = [](drakvuf_t drakvuf, mmvad_info_t* mmvad, void* callback_data)
         {
-            auto vctx = static_cast<visitor_context*>(ctx);
-            auto info = vctx->info;
-            auto plugin = vctx->plugin;
-            std::string name{(const char*)module_info->base_name->contents};
-            if (name.find("ntdll.dll") != std::string::npos)
+            if (!mmvad->file_name_ptr)
+                return false;
+
+            if (drakvuf_mmvad_type(drakvuf, mmvad) != 2) // VAD_TYPE_DLL
+                return false;
+
+            if (auto name = drakvuf_read_unicode(drakvuf, mmvad->file_name_ptr); !name.empty())
             {
-                plugin->proc_ntdll_hooked[info->attached_proc_data.pid] = true;
-                vctx->dll_base = module_info->base_addr;
-                *need_stop = true;
+                if (auto sub = name.find_last_of("/\\"); sub != std::string::npos)
+                    name.erase(0, sub + 1);
+
+                auto* vctx = (visitor_context_t*)callback_data;
+
+                if (name.find(vctx->name) != std::string::npos)
```

Comment:
This comparison is not perfect:
```
name = "msi.dllhack.dll"
vctx->name = "msi.dll"
```
and
```
name = "not_a_ntdll.dll"
vctx->name = "ntdll.dll"
```
would be `true`

Reply:
Please open an issue for this to be tracked as the PR has been merged

Reply:
The new code is logically identical to the old one:
```
if (name.find("ntdll.dll") != std::string::npos)
```
so the note is about the refactoring and not this PR.

Reply:
#1702